tlvc/fs/block_dev.c

```diff
@@ -222,17 +222,17 @@ static int raw_blk_rw(struct inode *inode, register struct file *filp,
 
 		ebh->b_blocknr = filp->f_pos >> SECT_SIZE_BITS;
 		o_data = bh->b_data;		/* save the 'original' values */
-		o_seg = ebh->b_L2seg;
-		ebh->b_L2seg = current->t_regs.ds;
-		bh->b_data = (unsigned char *)buf;
+		o_seg = (seg_t)ebh->b_L2seg;	/* may be long (xms active) or int */
+		ebh->b_L2seg = (ramdesc_t)current->t_regs.ds;
```

Comment:
Same idea here. While not buggy, as 16 bits is copied to 32, the cast shouldn't be necessary. Did GCC output a warning here without the cast?

Reply:
Thanks, this is a great catch @ghaerri - a serious bug. When analysing it now, this will have to eventually crash the system when XMS is active. b_L2seg is being saved, then restored and the upper half will be gone when the buffer is returned to the free-list (XMS active). 

I tested this just now, reducing the # of XMS buffers to 30 to force the issue and sure enough, the system goes dead when reusing that buffer. The reason it hasn't been caught yet is probably a combination of non-use (I haven't used XMS much for quite some time) and the huge # of XMS buffers we normally allocate. Plus, if there was a hang, it would not be easily repeatable and REALLY hard to catch.

Much appreciated.